src/common/types/types.cpp

```diff
@@ -958,5 +1028,324 @@ std::unique_ptr<LogicalType> LogicalType::ARRAY(std::unique_ptr<LogicalType> chi
+            throw ConversionException("Union casting is not supported");
+            // return tryCombineUnionTypes(left, right, result);
+        default:
+            KU_UNREACHABLE;
+            // LCOV_EXCL_END
+        }
+    } else {
+        LogicalTypeID resultID;
+        if (!tryGetMaxLogicalTypeID(left.typeID, right.typeID, resultID)) {
+            return false;
+        }
+        result = LogicalType(resultID);
+        return true;
+    }
+}
+
+LogicalTypeID LogicalTypeUtils::getMaxLogicalTypeID(const LogicalTypeID& left,
+    const LogicalTypeID& right) {
+    LogicalTypeID result;
+    if (!tryGetMaxLogicalTypeID(left, right, result)) {
+        throw common::BinderException(stringFormat("Cannot combine logical types {} and {}",
+            toString(left), toString(right)));
+    }
+    return result;
+}
+
+LogicalType LogicalTypeUtils::getMaxLogicalType(const LogicalType& left, const LogicalType& right) {
+    LogicalType result;
+    if (!tryGetMaxLogicalType(left, right, result)) {
+        throw common::BinderException(stringFormat("Cannot combine logical types {} and {}",
+            left.toString(), right.toString()));
+    }
+    return result;
+}
+
+// if we can combine the child types, then we can combine the list
+bool LogicalTypeUtils::tryCombineListTypes(const LogicalType& left, const LogicalType& right,
+    LogicalType& result) {
+    LogicalType childType;
+    if (!LogicalTypeUtils::tryGetMaxLogicalType(*ListType::getChildType(&left),
+            *ListType::getChildType(&right), childType)) {
+        return false;
+    }
+    result = *LogicalType::LIST(childType);
+    return true;
+}
+
+bool LogicalTypeUtils::tryCombineArrayTypes(const LogicalType& left, const LogicalType& right,
+    LogicalType& result) {
+    if (ArrayType::getNumElements(&left) != ArrayType::getNumElements(&right)) {
+        return tryCombineListTypes(left, right, result);
+    }
+    LogicalType childType;
+    if (!LogicalTypeUtils::tryGetMaxLogicalType(*ArrayType::getChildType(&left),
+            *ArrayType::getChildType(&right), childType)) {
+        return false;
+    }
+    result = *LogicalType::ARRAY(childType, ArrayType::getNumElements(&left));
+    return true;
+}
+
+// if we can match child labels and combine their types, then we can combine
+// the struct
+bool LogicalTypeUtils::tryCombineStructTypes(const LogicalType& left, const LogicalType& right,
+    LogicalType& result) {
+    auto leftFields = StructType::getFields(&left), rightFields = StructType::getFields(&right);
+    if (leftFields.size() != rightFields.size()) {
+        return false;
+    }
+    /*
+    auto comp = [](const StructField* const& a, const StructField* const& b) -> bool {
+        return a->getName() < b->getName();
+    };
+    std::sort(leftFields.begin(), leftFields.end(), comp);
+    std::sort(rightFields.begin(), rightFields.end(), comp);
+    // in general, kuzu seems to treat different element orderings as different types alltogether.
```

Comment:
Is this the case that we have discussed? If so, remove the comment. Otherwise let's discuss this part.